source/manual/intro-to-docker-even-more.html.md

```diff
@@ -0,0 +1,398 @@
+---
+title: Intro to Docker (Even More)
+section: Docker
+layout: manual_layout
+parent: "/manual.html"
+last_reviewed_on: 2019-11-06
+review_in: 6 months
+owner_slack: "#govuk-developers"
+---
+
+In the [Intro to Docker tutorial](/manual/intro-to-docker.html) we began with a generic container running the `ruby` image, and finished with a powerful `docker-compose` command to run the [content-publisher][] tests against a Postgres DB.
+
+This tutorial is going to pick up where we left off and introduce some more advanced concepts that we make use of in [govuk-docker][]. You should be able to understand the bulk of the repo after completing this tutorial.
+
+> You will need the following from the [previous tutorial](/manual/intro-to-docker.html):
+>
+>   * The `docker-compose.yml` file from Step 6
+>   * The Dockerfile that we wrote incrementally
+>   * The `.dockerignore` file from Step 3 (optional)
+
+## Step 1: Makefiles
+
+We did some tidy-up work at the end of the last tutorial, but there are still some manual commands to run.
+
+```shell
+$mac docker-compose run content-publisher-demo bundle install
+$mac docker-compose run content-publisher-demo bundle exec rake db:setup
+```
+
+To avoid having to remember these commands and type them for different projects, we can use a [Makefile][makefile].
+
+```
+content-publisher:
+  docker-compose run content-publisher-demo bundle install
+  docker-compose run content-publisher-demo bundle exec rake db:setup
+```
+
+> Makefiles use strict tabs for indentation, so you may need to adjust your editor if its inserting spaces automatically.
+
+Now we can run `make content-publisher` and the commands are run for us.
+
+## Step 2: `/govuk`
+
+At the end of the last tutorial we got some, but not all, of the tests passing. The remaining failures should be due to a missing dependency: [govuk-content-schemas][]. The tests are looking for the shemas at `../govuk-content-schemas`, which doesn't exist inside our container, since we only mapped the `content-publisher` directory to `/app`.
+
+* Mount the whole of the `~/govuk` directory
+
+```
+volumes:
+  - bundle:/usr/local/bundle
+  - ~/govuk:/govuk
+```
+
+* Adjust the working directory accordingly
+
+```
+working_dir: /govuk/content-publisher
+```
+
+* Check all of the tests now pass (hopefully!)
+
+```shell
+$mac docker-compose run content-publisher-demo bundle exec rake
```

Comment:
I was a little surprised to come across this syntax again, now that you've encapsulated these types of commands in the Makefile. Then I noticed this command is slightly different to the ones in the Makefile and that therefore we must run it verbosely like this.

May be worth saying that the Makefile is only for one-off setup?

Reply:
I think I see what you mean: we're only using `make` for setup, and not as a wrapper for running commands. I've added the word "setup" into a couple of places in the previous section. Does that help?